primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+
+        Calls the method `_unroll`, which does the actual work.
+
+        evidence : dict, optional
+            Dictionary containing the given evidence. The keys represent the
+            variable names of the given evidence and the values can either
+            be simple strings of the given values, or a np.array specifying
+            the strength of the evidence for each outcome, e.g.: For the binary
+            evidence node E `{"E": "True"}` is equivalent to
+            `{"E": np.array([1.0, 0.0])}`.
+        **kwargs
+            Takes the same keyword arguments as `set_evidence` and
+            `_set_evidence`.
+        """
+        if not self._evidence or evidence is not None:
+            self._set_evidence(evidence, **kwargs)
+        r = self._unroll()
+        self._evidence = False
+        return r
+
+    @abc.abstractmethod
+    def _set_evidence(self, evidence=None, **kwargs):
+        """Set evidence for the network (doing the actual work). (abstract)
+
+        See `set_evidence`.
+        """
+        pass
+
+    @abc.abstractmethod
+    def _unroll(self):
+        """Unroll the network (doing the actual work). (abstract)
+
+        See `unroll`.
+        """
+        pass
+
+    @property
+    def t(self):
+        """Get number current time-slice.
+
+        Returns
+        -------
+        int
+            The number of the current time-slice.
+        """
+        return self._t
+
+
+class PriorFeedbackExact(DBNInferenceMethod):
+    """Prior feedback-based approach to unrolling a DBN, using exact inference.
+
+    Inference is exact and based on factor trees. The marginal posterior
+    probabilities of the variables of the current time-slice are used as prior
+    probabilties for the next time-slice. They are set as the conditional
+    probability distribution of the nodes/variables.
+    """
+
+    def __init__(self, dbn):
+        super(PriorFeedbackExact, self).__init__(dbn)
+        self._dbn = copy.deepcopy(dbn)
+        self._ft = None
+
+    def marginal_probabilities(self, variable):
+        """Get the marginal probabilities of a variable.
+
+        See `DBNInferenceMethod.marginal_probabilities`.
+        """ 
+        f = self.marginals([variable])
+        return dict(zip(f.values[variable], list(f.get_potential())))
+
+    def marginals(self, variables):
+        """Get the marginals of a variable.
+
+        Parameters
+        ----------
+        variables : [String,]
+            List containing the names of the variables whose joint prior
+            or posterior marginals are desired.
+            Joint marginals for multiple variables can only be computed
+            if they are contained in any clique.     
+                
+        Returns
+        -------
+        Factor
+            A factor containing the desired marginals.
+        """
+        return self._ft.marginals(variables)
+
+    def _set_evidence(self, evidence=None, soft_posteriors=False, **kwargs):
```

Comment:
Do you need to still pass the kwargs here? If abstractmethod does not force you to, I would suggest removing them here, since they are neither documented nor used.